icepack/utilities.py

```diff
@@ -46,21 +46,56 @@ def get_kwargs_alt(dictionary, keys, keys_alt):
                   for key, alt_key in zip(keys, keys_alt)))
 
 
-def facet_normal_2(mesh):
+def get_mesh_dimensions(mesh):
+    r"""Get the number of dimensions in the mesh. Half dimensions for extruded into the vertical.
+    """
+    dim = mesh.geometric_dimension()
+    if mesh.layers is not None:
+        dim -= 0.5
+    return dim
+
+def facet_normal_nd(mesh):
     r"""Compute the horizontal component of the unit outward normal vector
     to a mesh"""
-    ν = firedrake.FacetNormal(mesh)
-    return firedrake.as_vector((ν[0], ν[1]))
+    dim = get_mesh_dimensions(mesh)
+    if dim == 2:
+        return firedrake.FacetNormal(mesh)
+    elif dim == 2.5:
+        ν = firedrake.FacetNormal(mesh)
+        return firedrake.as_vector((ν[0], ν[1]))
+    elif dim in [1, 1.5]:
+        return firedrake.FacetNormal(mesh)[0]
 
 
-def grad_2(q):
+def grad_nd(q):
     r"""Compute the horizontal gradient of a 3D field"""
-    return firedrake.as_tensor((q.dx(0), q.dx(1)))
+    dim = get_mesh_dimensions(q.ufl_domain())
+    if dim == 2:
+        return firedrake.grad(q)
+    elif dim == 2.5:
+        return firedrake.as_tensor((q.dx(0), q.dx(1)))
+    elif dim in [1, 1.5]:
+        return q.dx(0)
+
+
+def ds_nd(q):
```

Comment:
You can get rid of this helper function entirely (see my comment on the mass transport solver).

Reply:
I kept it actually because you need it in two places (mass_transport and flow_solver) so it is a bit cleaner to leave it here. If you want to take it out though that is fine.